src/Database/Redis/Cluster.hs

```diff
@@ -146,12 +149,18 @@ instance Exception NoNodeException
 data TimeoutException = TimeoutException String deriving (Show, Typeable)
 instance Exception TimeoutException
 
-connect :: (Host -> CC.PortID -> Maybe Int -> IO CC.ConnectionContext) -> [CMD.CommandInfo] -> MVar ShardMap -> Bool -> ([NodeConnection] -> IO ShardMap) -> TcpInfo -> IO Connection
+data RefreshNeededException = NodeMissing [B.ByteString]
+    | AllNodeMissing
+    | RequestTimingOut [[B.ByteString]]
+    deriving (Show, Typeable)
+instance Exception RefreshNeededException
+
+connect :: (Host -> CC.PortID -> Maybe Int -> IO CC.ConnectionContext) -> [CMD.CommandInfo] -> TMVar (ShardMap, Time.UTCTime) -> Bool -> ([NodeConnection] -> IO ShardMap) -> TcpInfo -> IO Connection
 connect withAuth commandInfos shardMapVar isReadOnly refreshShardMap (tcpInfo@TcpInfo{ timeoutOpt, maxResources, idleTime }) = do
-        shardMap <- readMVar shardMapVar
+        shardMap <- atomically $ readTMVar shardMapVar
```

Comment:
Ideally we should aim for not blocking all threads using Atomically here. We need to do a lock free read on shardMap otherwise this can be a perf bottleneck.

Reply:
I am not sure this understanding is right. But @shashitnak you can use `IO` version of functions like `readTMVarIO` rather than atomically to convert from STM to IO.